### PR TITLE
Issue_660: Fixed NRFU1.2 NtpAssocitionsTests - Tests incorrectly failing

### PR DESCRIPTION
**Pull request by geetanjalimanegslab**
# Please include a summary of the changes
- Updated nrfu_tests/master_def.yaml with NRFU1.2 testcase changes
- Updated nrfu_tests/test_base_services_ntp_association.py  with the support for single ntp server
- Updated nrfu_tests/test_definition.yaml with NRFU1.2 testcase changes
-Updated nrfu_tests/test_definition.yaml.j2 with NRFU1.2 testcase changes

# Any specific logic/part of code you need extra attention on

Explain any specific logic you need special attention on

# Include the Issue number and link
#660 
Make sure to link the issue in the github PR UI 

# List/Attach any dependencies/past issues that are required for this change/provide context

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [x] New feature (non-breaking change which adds functionality)
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [ ] Other (please specify)

# Effort required on reviewers end

- - [ ] Easy
- - [x] Medium
- - [ ] Hard 

# How Has This Been Tested?

## Bug fix

    Include before and after snapshots/screenshots/changed logs
    
## New feature

    Ensure current test cases pass and include newer test cases if required
    
# CI pipeline result

- - [ ] Pass
- - [ ] Fail
  
  If it fails, explain the reason and whether or not we should ignore the failure
  
# Verify Documentation Update

    If applicable, ensure documentation has been updated for ReadMe, Getting Started guide, and Style guide
    
# Additional comments
Reports:
- Tested with single NTP server check is True and the secondary server is not found: Passed (Tested with hardcoded output)
[report_pass_ntp_single_check_True_sec_association_not_found.docx](https://github.com/aristanetworks/vane/files/14663935/report_pass_ntp_single_check_True_sec_association_not_found.docx)

- Tested with single NTP server check is True: Passed
  [report_pass_ntp_single_check_True.docx](https://github.com/aristanetworks/vane/files/14663937/report_pass_ntp_single_check_True.docx)

- Tested with a single NTP server check is False: Passed
  [report_pass_ntp_single_check_false.docx](https://github.com/aristanetworks/vane/files/14663939/report_pass_ntp_single_check_false.docx)

-  Tested with a single NTP server check is False, and secondary ntp server association is not present: Failed  (Tested with hardcoded output)
    [report_fail_ntp_single_check_false_sec_association_not_found.docx](https://github.com/aristanetworks/vane/files/14663943/report_fail_ntp_single_check_false_sec_association_not_found.docx)

- Tested with NTP server is not configured: Skipped
  [report_ntp_skipped.docx](https://github.com/aristanetworks/vane/files/14663947/report_ntp_skipped.docx)

- Tested with NTP server association details  not found: Failed
  [report_ntp_servr_association_not_found.docx](https://github.com/aristanetworks/vane/files/14663960/report_ntp_servr_association_not_found.docx)

**HTML Reports**
[NRFU_1_2.zip](https://github.com/aristanetworks/vane/files/14663963/NRFU_1_2.zip)

### CODE REVIEW

File: nrfu_tests/test_base_services_ntp_association.py [hunk 1]
```diff
@@ -31,15 +31,28 @@ def test_ntp_clocks(self, dut, tests_definitions):
...
             "primary_ntp_association": "Not found",
             "secondary_ntp_association": "Not found",
         }
-        # Forming an output message if a test result is pass
+        single_ntp_check = test_params["input"].get("single_ntp_check")
         tops.output_msg = (
-            "NTP server is configured on the device. Primary and secondary NTP association is"
-            " correct on the device."
+            "NTP server should be configured on the device. Primary and secondary NTP association"
+            " is correct on the device."
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_base_services_ntp_association.py:42]:
```suggestion
            "Primary and secondary NTP association is correct on the device."
```

File: nrfu_tests/test_base_services_ntp_association.py [hunk 2]
```diff
@@ -53,49 +66,56 @@ def test_ntp_clocks(self, dut, tests_definitions):
...
+                )
+                tests_tools.post_process_skip(tops, self.test_ntp_clocks, self.output)
+                pytest.skip(tops.output_msg)
 
             ntp_association_cmd = "show ntp associations"
 
             """
             TS: Running `show ntp associations` on device and verifying that the
-            primary and secondary NTP association is correct on the host.
+            expected NTP association is correct on the host.
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_base_services_ntp_association.py:80]:
```suggestion
            NTP association is correct on the host.
```

File: nrfu_tests/test_base_services_ntp_association.py [hunk 3]
```diff
@@ -53,49 +66,56 @@ def test_ntp_clocks(self, dut, tests_definitions):
...
-                    tops.output_msg += (
-                        "Primary and Secondary NTP associations are not found on the device.\n"
-                    )
-                elif not_primary_ntp_association:
+                not_primary_ntp_association = tops.actual_output.get(
+                    "primary_ntp_association"
+                ) != tops.expected_output.get("primary_ntp_association")
+                not_secondary_ntp_association = tops.actual_output.get(
+                    "secondary_ntp_association"
+                ) != tops.expected_output.get("secondary_ntp_association")
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_base_services_ntp_association.py:107]: Can we write it as below to make it more readable.
```suggestion
                actutal_primary_ntp_status = tops.actual_output.get("primary_ntp_association")
                expected_primary_ntp_status = tops.expected_output.get("primary_ntp_association")
                actual_secondary_ntp_status =  tops.actual_output.get("secondary_ntp_association")
                expected_secondary_ntp_status =  tops.expected_output.get("secondary_ntp_association")
                
                no_primary_ntp = actutal_primary_ntp_status != expected_primary_ntp_status
                no_secondary_ntp = actual_secondary_ntp_status  != expected_secondary_ntp_status 
```

File: nrfu_tests/test_base_services_ntp_association.py [hunk 4]
```diff
@@ -53,49 +66,56 @@ def test_ntp_clocks(self, dut, tests_definitions):
...
+                if not_primary_ntp_association:
                     tops.output_msg += "Primary NTP association is not found on the device.\n"
-                elif not_secondary_ntp_association:
-                    tops.output_msg += "Secondary NTP association is not found on the device."
+                elif not single_ntp_check:
+                    if not_secondary_ntp_association:
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_base_services_ntp_association.py:111]:
```suggestion
                elif not single_ntp_check and not_secondary_ntp_association:
```

Can we do it as .

File: nrfu_tests/test_definition.yaml.j2 [hunk 5]
```diff
@@ -16,12 +16,15 @@
     - name: test_ntp_clocks
       description: Test case to verify that 2 NTP clocks(a peer and a candidate) are locked on the device.
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_definition.yaml.j2:17]: Do we need to update the test case description now ?

File: nrfu_tests/test_base_services_ntp_association.py [hunk 6]
```diff
@@ -53,49 +66,56 @@ def test_ntp_clocks(self, dut, tests_definitions):
...
             output = tops.run_show_cmds([ntp_association_cmd])
             logging.info(
                 f"On device {tops.dut_name} output of {ntp_association_cmd} command is:\n{output}\n"
             )
             self.output += f"Output of {ntp_association_cmd} command is: \n{output}"
             ntp_associations = output[0]["result"].get("peers")
-            assert ntp_associations, "NTP association details are not collected."
+            assert (
+                ntp_associations
+            ), "NTP association details are not collected in the command output."
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_base_services_ntp_association.py:90]:
```suggestion
            ), f"No NTP servers found in '{ntp_association_cmd}' command output."
```

File: nrfu_tests/test_base_services_ntp_association.py [hunk 7]
```diff
@@ -31,71 +31,93 @@ def test_ntp_clocks(self, dut, tests_definitions):
...
+                )
+                tests_tools.post_process_skip(tops, self.test_ntp_clocks, self.output)
+                pytest.skip(tops.output_msg)
 
             ntp_association_cmd = "show ntp associations"
 
             """
-            TS: Running `show ntp associations` on device and verifying that the
-            primary and secondary NTP association is correct on the host.
+            TS: Running `show ntp associations` command on the device and verifying that the
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_base_services_ntp_association.py:79]: Remove white space.
image

**Review by vishwajeetgslab — APPROVED** [review]

**Comment by kgrozis-arista** [comment]
@vishwajeetgslab 
I made a mistake and reviewed this PR in another PR.  Moving thread to correct PR.
https://github.com/aristanetworks/vane/pull/704#issuecomment-2011681114

**Comment by kgrozis-arista** [comment]
> @vishwajeetgslab I made a mistake and reviewed this PR in another PR. Moving thread to correct PR. [#704 (comment)](https://github.com/aristanetworks/vane/pull/704#issuecomment-2011681114)

Questions from previous thread:
Hi @kgrozis-arista ,

We tried to reproduce above issue(with issue_660 branch), however we are not able to reproduce it. Could you please help us with following:

Can you please share the branch, you tried to run this test case?
Can you please share the command added in test_definitions.yaml ? our primary investigation is that command added in the test_definitions.yaml might be incorrect hence vane is returning a string for that command.
if possible could you please share the report?

**Comment by kgrozis-arista** [comment]
> > @vishwajeetgslab I made a mistake and reviewed this PR in another PR. Moving thread to correct PR. [#704 (comment)](https://github.com/aristanetworks/vane/pull/704#issuecomment-2011681114)
> 
> Questions from previous thread: Hi @kgrozis-arista ,
> 
> We tried to reproduce above issue(with issue_660 branch), however we are not able to reproduce it. Could you please help us with following:
> 
> Can you please share the branch, you tried to run this test case?
I'm using the branch from the issue:
```
(vane-py3.9) kgrozis@Kevins-MacBook-Pro vane % git status
On branch 660-nrfu12-ntpassocitionstests-tests-incorrectly-failing
Your branch is up to date with 'origin/660-nrfu12-ntpassocitionstests-tests-incorrectly-failing'.
```
 Can you please share the command added in test_definitions.yaml ?
 Zip includes test_definition and repots 
[nrfu1.2.zip](https://github.com/aristanetworks/vane/files/14695893/nrfu1.2.zip)


 our primary investigation is that command added in the test_definitions.yaml might be incorrect hence vane is returning a string for that command. if possible could you please share the report?
 Zip includes test_definition and repots

**Comment by vishwajeetgslab** [comment]
> > > @vishwajeetgslab I made a mistake and reviewed this PR in another PR. Moving thread to correct PR. [#704 (comment)](https://github.com/aristanetworks/vane/pull/704#issuecomment-2011681114)
> > 
> > 
> > Questions from previous thread: Hi @kgrozis-arista ,
> > We tried to reproduce above issue(with issue_660 branch), however we are not able to reproduce it. Could you please help us with following:
> > Can you please share the branch, you tried to run this test case?
> > I'm using the branch from the issue:
> 
> ```
> (vane-py3.9) kgrozis@Kevins-MacBook-Pro vane % git status
> On branch 660-nrfu12-ntpassocitionstests-tests-incorrectly-failing
> Your branch is up to date with 'origin/660-nrfu12-ntpassocitionstests-tests-incorrectly-failing'.
> ```
> 
> Can you please share the command added in test_definitions.yaml ? Zip includes test_definition and repots [nrfu1.2.zip](https://github.com/aristanetworks/vane/files/14695893/nrfu1.2.zip)
> 
> our primary investigation is that command added in the test_definitions.yaml might be incorrect hence vane is returning a string for that command. if possible could you please share the report? Zip includes test_definition and repots

Hi @kgrozis-arista ,
The issue seems that `show ntp status` command output is not collected as i can see same in the logger and when code tries to fetch the output.get("status") it fails.

image

I switched to your branch and tested the same against the hardware lab however i was not able to reproduce it.

**Comment by kgrozis-arista** [comment]
> > > > @vishwajeetgslab I made a mistake and reviewed this PR in another PR. Moving thread to correct PR. [#704 (comment)](https://github.com/aristanetworks/vane/pull/704#issuecomment-2011681114)
> > > 
> > > 
> > > Questions from previous thread: Hi @kgrozis-arista ,
> > > We tried to reproduce above issue(with issue_660 branch), however we are not able to reproduce it. Could you please help us with following:
> > > Can you please share the branch, you tried to run this test case?
> > > I'm using the branch from the issue:
> > 
> > 
> > ```
> > (vane-py3.9) kgrozis@Kevins-MacBook-Pro vane % git status
> > On branch 660-nrfu12-ntpassocitionstests-tests-incorrectly-failing
> > Your branch is up to date with 'origin/660-nrfu12-ntpassocitionstests-tests-incorrectly-failing'.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can you please share the command added in test_definitions.yaml ? Zip includes test_definition and repots [nrfu1.2.zip](https://github.com/aristanetworks/vane/files/14695893/nrfu1.2.zip)
> > our primary investigation is that command added in the test_definitions.yaml might be incorrect hence vane is returning a string for that command. if possible could you please share the report? Zip includes test_definition and repots
> 
> Hi @kgrozis-arista , The issue seems that `show ntp status` command output is not collected as i can see same in the logger and when code tries to fetch the output.get("status") it fails.
> 
> image
> I switched to your branch and tested the same against the hardware lab however i was not able to reproduce it.

@vishwajeetgslab I'm double check my systems and make sure nothing is incorrect on my side

**Comment by kgrozis-arista** [comment]
> > > > > @vishwajeetgslab I made a mistake and reviewed this PR in another PR. Moving thread to correct PR. [#704 (comment)](https://github.com/aristanetworks/vane/pull/704#issuecomment-2011681114)
> > > > 
> > > > 
> > > > Questions from previous thread: Hi @kgrozis-arista ,
> > > > We tried to reproduce above issue(with issue_660 branch), however we are not able to reproduce it. Could you please help us with following:
> > > > Can you please share the branch, you tried to run this test case?
> > > > I'm using the branch from the issue:
> > > 
> > > 
> > > ```
> > > (vane-py3.9) kgrozis@Kevins-MacBook-Pro vane % git status
> > > On branch 660-nrfu12-ntpassocitionstests-tests-incorrectly-failing
> > > Your branch is up to date with 'origin/660-nrfu12-ntpassocitionstests-tests-incorrectly-failing'.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Can you please share the command added in test_definitions.yaml ? Zip includes test_definition and repots [nrfu1.2.zip](https://github.com/aristanetworks/vane/files/14695893/nrfu1.2.zip)
> > > our primary investigation is that command added in the test_definitions.yaml might be incorrect hence vane is returning a string for that command. if possible could you please share the report? Zip includes test_definition and repots
> > 
> > 
> > Hi @kgrozis-arista , The issue seems that `show ntp status` command output is not collected as i can see same in the logger and when code tries to fetch the output.get("status") it fails.
> > image
> > I switched to your branch and tested the same against the hardware lab however i was not able to reproduce it.
> 
> @vishwajeetgslab I'm double check my systems and make sure nothing is incorrect on my side

@vishwajeetgslab 
I found the issue.  The hardware I was using has an early version of EOS (4.23).  This did not support json output for `show ntp status`:
```
mn413...14:57:05#sh ntp status | json
% This is an unconverted command
{
    "errors": [
        "This is an unconverted command"
    ]
}
```
When I use a system with newer EOS, the test case works correctly.  I'll finish the review and update you with any comments.

**Review by kgrozis-arista — APPROVED** [review]
LGTM

**Pull request merged**